java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -0,0 +1,65 @@
+/** Provides classes to reason about XPath vulnerabilities. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * An abstract type representing a call to interpret XPath expressions.
+ */
+class XPathSink extends MethodAccess {
+  /**
+   * Gets the argument representing the XPath expressions to be evaluated.
+   */
+  abstract Expr getSink();
+}
+
+/** The class `javax.xml.xpath.XPath` */
+class XPath extends RefType {
+  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
+}
+
+/** A call to `XPath.evaluate` or `XPath.compile` */
+class XPathEvaluateOrCompile extends XPathSink {
+  XPathEvaluateOrCompile() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
+      m.hasName(["evaluate", "compile"])
+    )
+  }
+
+  override Expr getSink() { result = this.getArgument(0) }
+}
+
+/** Any class extending or implementing `org.dom4j.Node` */
+class Dom4JNode extends RefType {
+  Dom4JNode() {
+    exists(Interface node | node.hasQualifiedName("org.dom4j", "Node") |
+      this.extendsOrImplements*(node)
+    )
+  }
+}
+
+/** A call to `Node.selectNodes` or `Node.selectSingleNode` */
+class NodeSelectNodes extends XPathSink {
+  NodeSelectNodes() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof Dom4JNode |
+      m.hasName(["selectNodes", "selectSingleNode"])
+    )
+  }
+
+  override Expr getSink() { result = this.getArgument(0) }
+}
+
+/** A sink that represents a method that interprets XPath expressions. */
+class XPathInjectionSink extends DataFlow::ExprNode {
```

Comment:
We probably want to make this class an extension point where customers will be able to add their own XPath injection sinks. To do this, the class should be made `abstract` (and extend `Node` rather than `ExprNode` in order to not make restrict it prematurely), and the sinks defined here should be a `private` subclass (`DefaultXPathInjectionSink` would be a good name for such a subclass).

Reply:
Yes, I tried to make `XPathSink` the extension point for that same reason, but I can see how it was redundant (as you pointed out in https://github.com/github/codeql/pull/5774#discussion_r620949790). 

Followed your advice in 2054791, thanks for the feedback!